rotkehlchen/chain/evm/decoding/odos/v2/decoder.py

```diff
@@ -0,0 +1,241 @@
+from rotkehlchen.constants.resolver import evm_address_to_identifier
+from rotkehlchen.errors.misc import RemoteError
+from rotkehlchen.errors.serialization import DeserializationError
+from rotkehlchen.fval import FVal
+from rotkehlchen.history.events.structures.evm_event import EvmEvent
+from rotkehlchen.history.events.structures.types import HistoryEventSubType, HistoryEventType
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTokenKind
+from rotkehlchen.utils.misc import hex_or_bytes_to_address, hex_or_bytes_to_int
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.user_messages import MessagesAggregator
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class Odosv2DecoderBase(DecoderInterface):
+    def __init__(
+            self,
+            evm_inquirer: 'EvmNodeInquirer',
+            base_tools: 'BaseDecoderTools',
+            msg_aggregator: 'MessagesAggregator',
+            router_address: ChecksumEvmAddress,
+            native_currency: 'Asset',
+    ) -> None:
+        super().__init__(
+            evm_inquirer=evm_inquirer,
+            base_tools=base_tools,
+            msg_aggregator=msg_aggregator,
+        )
+        self.evm_txns = EvmTransactions(self.evm_inquirer, self.base.database)
+        self.router_address = router_address
+        self.native_currency = native_currency.resolve_to_crypto_asset()
+
+    def _resolve_tokens_data(
+            self,
+            token_addresses: list['ChecksumEvmAddress'],
+            token_amounts: list[int],
+    ) -> dict[str, FVal]:
+        """Returns the resolved evm tokens or native currency and their amounts"""
+        resolved_result: dict[str, FVal] = {}
+        for token_address, token_amount in zip(token_addresses, token_amounts, strict=True):
+            if token_address == ZERO_ADDRESS:
+                asset = self.native_currency
+            else:
+                asset = get_or_create_evm_token(
+                    userdb=self.base.database,
+                    evm_inquirer=self.base.evm_inquirer,
+                    evm_address=token_address,
+                    chain_id=self.base.evm_inquirer.chain_id,
+                    token_kind=EvmTokenKind.ERC20,
+                )
+
+            resolved_result[asset.identifier] = asset_normalized_value(amount=token_amount, asset=asset)  # noqa: E501
+
+        return resolved_result
+
+    def _calculate_router_fee(
+            self, context: 'DecoderContext',
+            sender: 'ChecksumEvmAddress',
+            output_tokens: dict[str, FVal],
+    ) -> list[EvmEvent]:
+        """Calculates the fees collected by the router and returns the fee events.
+        Fees are the amount of out tokens that router received - the amount that router sent
+        """
```

Comment:
can it raise remoteError if for some reason it has to query for the internal transactions?

Reply:
no because they are handled where the internal transactions are queried